File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    answer = []
    

    
    def dfs(L, now, ans):
        if L == len(tickets):
            answer.append(ans)
            return
        
        for i in range(len(tickets)):
            if not visited[i] and tickets[i][0] == now:
                visited[i] = 1
                dfs(L+1, tickets[i][1], ans + " "+ tickets[i][1])
                visited[i] = 0
        
    visited = [0]*len(tickets)
    dfs(0,"ICN", "ICN")
        
    answer.sort()

    
    return answer[0].split(" ")

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
class Solution {
    static int n;
    static boolean[] visited;
    static ArrayList<String> arr;
    static String[][] ticket;
    
    List<String> answer;
    public String[] solution(String[][] tickets) {
        
        n = tickets.length;
        visited = new boolean[n];
        ticket = tickets;
        
        
        Arrays.sort(ticket, (a,b) -> a[1].compareTo(b[1]));
        arr = new ArrayList<>();
        
        
        dfs(0, "ICN", "ICN");
        
        
        
        String[] answer = arr.get(0).split(",");
        
        return answer;
    }
    static void dfs(int L, String target, String path){
        if(L==n){
            arr.add(path);
            return;
        }
        
        for(int i=0; i<n; i++){
            if(!visited[i]&&target.equals(ticket[i][0])){
                visited[i] = true;
                dfs(L+1, ticket[i][1], path + ","+ticket[i][1] );
                visited[i] = false;
            }
        }
    }
}